### DISCUSSION TITLE
Handling "Stream error in the HTTP/2 framing layer on handle ID"

### DISCUSSION THREAD

**Discussion by elhennig** [abraunegg/onedrive#2698] [General]
Hi,
since this week my onedrive sync fails uploading files with the error

>  "Stream error in the HTTP/2 framing layer on handle ID"

he solution or better  workaround seems to be the option 

> force_http_11 = "true"

in the config. Is there any other "real" solution to make the uploads work also with HTTP 2 again?

The versions installed on my opensuse leap 15.5 system are:
- libcurl4 8.0.1-150400.5.44.1 
- onedrive 2.4.25-bp155.2.3.1

**Comment by abraunegg** [comment]
The error message "Error in the HTTP2 framing layer on handle <handle_ID>" suggests an issue within the HTTP/2 protocol layer, particularly with how frames are being processed or interpreted. HTTP/2 introduces a binary framing layer at its core, which is a departure from the newline-delimited text format of HTTP/1.x. This binary framing layer is designed to improve performance over HTTP/1.x by allowing multiplexing of requests and responses, stream prioritisation, and more efficient use of network resources. However, this complexity also introduces new areas for potential errors.

There are several possible reasons for a HTTP2 framing error:

1. **Protocol Violations:** The error could be due to a protocol violation, where the frames sent by the client (onedrive via libcurl) or server (OneDrive API) do not conform to the specifications defined by the HTTP/2 standard. This could include incorrect frame sizes, invalid frame sequences, or use of reserved bit fields in ways that are not allowed. I doubt this is the problem as if this was the issue - this would be way more widespread and the blast radius of impact would be huge (reported everywhere, where HTTP2 is being used with Microsoft OneDrive).

2. **Network Issues:** Underlying network issues such as packet loss, corruption, or transmission errors can indirectly cause framing errors if they lead to malformed frames being received. Please check your network for packet loss between you and the Microsoft OneDrive API service. It is beyond this discussion / issue ticket to assist you to do this. 

3. **Interoperability Issues:** The error might arise from interoperability issues between different implementations of the HTTP/2 protocol. For example, if there is a transparent upstream device from you (such as a proxy or load balancer or firewall) or the server (the OneDrive API) has bugs or handles certain HTTP/2 frames differently, it might lead to errors. I doubt this is the OneDrive API here, however I have no idea of your environment path to the OneDrive API service - so you may have some sort of transparent upstream device breaking things for you.

4. **Libcurl Bugs:** There could be bugs in the libcurl version you are using or in the server handling the HTTP/2 connections. 

 As per the above, your only *real* option with your current platform and environment is:
 
 * Enable the following options on your system:
   * Tell the client to use HTTP/1.1 via `force_http_11 = "true"`
   * Tell the client to use IPv4 only via `ip_protocol_version = "1"`

> Is there any other "real" solution to make the uploads work also with HTTP 2 again?

You may consider the following options:

1. Update your platform to something more modern
2. Check for upstream path for any firewall, proxy or other device that is performing traffic filtering, redirection, analysis and so on. You need to identify what has changed in your environment between you and Microsoft OneDrive that is potentially causing this issue.
3. Report to your distribution support channel that 'curl' is breaking in HTTP/2 protocol handling (raise a bug with them), and get the distribution to provide you a newer version of 'curl' fixing any bug or issue that is causing 'Stream error in the HTTP/2 framing layer'
4. Compile 'curl' / 'libcurl' yourself - you might need to do this to assist in performing option 3 as evidence a newer version of curl resolves the issue.
